session/session.go

```diff
@@ -243,12 +230,14 @@ func (s *Session) run(ctx context.Context) {
 	s.periodicSearchTimer = time.NewTimer(s.periodicSearchDelay.NextWaitTime())
 	for {
 		select {
-		case blk := <-s.incoming:
-			if blk.counterMessage {
-				s.updateReceiveCounters(ctx, blk)
-			} else {
-				s.handleIncomingBlock(ctx, blk)
+		case rcv := <-s.incoming:
+			s.cancelIncomingBlocks(ctx, rcv)
```

Comment:
I'm concerned these may not be in the right order:

1. If we cancel before we update things, will we think all blocks are duplicates?
2. Will we even _accept_ them? That is, don't we have to call `handleIncomingBlocks` first?

Reply:
I believe these functions are mutually exclusive:
- `receiveBlocks()` only modifies the session's live wants, tofetch and pastWants CID lists.
- `cancelIncomingBlocks()` tells the latency tracker that the CID was cancelled, and tells the want manager to send out a cancel message

Reply:
I will move `cancelIncomingBlocks ()` below `handleIncomingBlocks()` though, that's probably clearer

Reply:
Actually `cancelIncomingBlocks()` must be called *before* `receiveBlocks()` because `receiveBlocks()` removes the CID from the want list: `cancelIncomingBlocks()` relies on `cidIsWanted()` returning true

Reply:
Note that I moved this code from the bottom of `ReceiveBlocksFrom()`, where it appeared that it was getting called after `handlingIncomingBlocks()`, but in fact it was being called *before* `handlingIncomingBlocks()` because the call to `handlingIncomingBlocks()` was asynchronous.

Reply:
Ah, ok, I'm just confusing myself. This happens _after_ storing the actual blocks.